torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -0,0 +1,125 @@
+    auto_wrap_policy = auto_wrap_kwargs["auto_wrap_policy"]
+    root_module = auto_wrap_kwargs["module"]
+    assert auto_wrap_policy is not None
+    # For auto wrapping, submodules should not already be wrapped with FSDP
+    # since double wrapping is not supported
+    for module_name, module in root_module.named_modules():
+        if isinstance(module, module_wrapper_cls):
+            raise ValueError(
+                f"Expected {module_name} to NOT be FullyShardedDataParallel "
+                "if using an `auto_wrap_policy`"
+            )
+    mixed_precision = fsdp_kwargs["mixed_precision"]
+    if mixed_precision is not None and _contains_batchnorm(root_module):
+        _override_batchnorm_mixed_precision(root_module)
+        auto_wrap_policy = functools.partial(
+            _or_policy, policies=[_wrap_batchnorm_individually, auto_wrap_policy]
+        )
+        warnings.warn(
+            "Both mixed precision and an `auto_wrap_policy` were specified "
+            "for FSDP, where the wrapped module has batch norm submodules. "
+            "The batch norm submodules will be wrapped as separate FSDP "
+            "instances with mixed precision disabled since some batch norm "
+            "kernels do not support low precision."
+        )
+        auto_wrap_kwargs["auto_wrap_policy"] = auto_wrap_policy
+    _recursive_wrap(**auto_wrap_kwargs, **fsdp_kwargs)
+
+
+def _get_params_per_wrapped_module(
+    root_module: nn.Module,
+    auto_wrap_policy: Callable,
+    ignored_modules: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+) -> List[List[nn.Parameter]]:
+    """
+    Returns the parameters per wrapped module according to the given auto wrap
+    policy and ignored modules/parameters without actually performing any
+    explicit module wrapping.
+    """
+    # Record the modules to wrap without actually wrapping
+    wrapped_modules: List[nn.Module] = []
+    wrapper_cls = functools.partial(_record_module_wrapper_cls, wrapped_modules)
+    _recursive_wrap(
+        root_module,
+        auto_wrap_policy=auto_wrap_policy,
+        wrapper_cls=wrapper_cls,
+        ignored_modules=ignored_modules,
+        ignored_params=ignored_params,
+        only_wrap_children=False,
+    )
+    # Always include the root module even if not wrapped by the given policy
+    if root_module not in wrapped_modules:
+        wrapped_modules.append(root_module)
+
+    params_per_wrapped_module: List[List[nn.Parameter]] = []
+    visited_params = set()
+    # Constructing `wrapped_modules` with `_recursive_wrap()` orders the
+    # modules following a post-order traversal. We record parameters in
+    # `params_per_wrapped_module` using a reverse post-ordering, which is a
+    # topological sort, so that each shared parameter is guaranteed to be
+    # grouped with its lowest common ancestor module's parameters.
+    wrapped_modules.reverse()
+    wrapped_modules_set = set(wrapped_modules)
+    for module_to_wrap in wrapped_modules:
```

Comment:
check my understanding. 

1. this for loop run in the reverse of post-order traversal and also starts from the root
2. the inner BFS basically finds all params not wrapped by any FSDP units, which should be taken care of by the current one just popped from the queue. 

> so that each shared parameter is guaranteed to be grouped with its lowest common ancestor module's parameters.

Is "lowest common ancestor module" refers to the lowest common FSDP ancestor module? If yes, then this part I didn't follow. Suppose we have a param p used by submodule A and B. A and B are wrapped by different FSDP unit. In this case, I assume p would still appear in both A's and B's flat param? And our suggestion is that user should wrap A and B in the same FSDP unit?

Reply:
Thanks for catching this! This reverse-post order traversal and BFS only handles parent-child shared parameters, not sibling shared parameters.

- Parent-child shared parameters: Submodule `A` and submodule `B` are both to be logically wrapped and share a parameter `p`, and `A` is the parent of `B`. The current implementation will make sure that `p` is assigned to `A`.
- Sibling shared parameters: Submodule `A` and submodule `B` are both to be logically wrapped and share a parameter `p`, and neither `A` is the parent of `B` nor `B` is the parent of `A`. We need to rewrite the logic to traverse the module tree and pre-compute the shared parameters. Then, for each shared parameter, we need to compute the least common ancestor submodule among the ones to be logically wrapped and manually assign the shared parameter to that LCA.

I will explore this in a follow-up.